### ISSUE TITLE
Higher version transitive doesn't align its parent dependency.

### ISSUE THREAD

**Issue by chali** [nebula-plugins/gradle-nebula-integration#52]
I found the following situation. `netflix:exampleapp-client:80.0.139` depends on `netflix:exampleapp-common:80.0.154`. Modules `exampleapp-*` are aligned. It causes upgrade of `netflix:exampleapp-client` to `80.0.154` which depends on a higher version of `common` and the situation repeats. With our resolution rules it converges to this situation:

```
compile - Dependencies for source set 'main' (deprecated, use 'implementation' instead).
\--- netflix:exampleapp-client:80.0.139 -> 80.0.236
     +--- netflix:exampleapp-common:80.0.260
     \--- netflix:exampleapp-smart-client:80.0.21
          \--- netflix:exampleapp-model:80.0.15
```

However, with `belongsTo` based implementation, it stops sooner. `netflix:exampleapp-common:80.0.249` is not causing an upgrade of `netflix:exampleapp-client` to the closest possible version which should be `80.0.236`

```
compile - Dependencies for source set 'main' (deprecated, use 'implementation' instead).
\--- netflix:exampleapp-client:80.0.139 -> 80.0.225
     +--- netflix:exampleapp-common:80.0.249
     \--- netflix:exampleapp-smart-client:80.0.10
```

Complete example: https://github.com/nebula-plugins/gradle-nebula-integration/tree/master/non-upgraded-dependency-by-alignment

You can comment out:
```
components.all { ComponentMetadataDetails details ->
    if (details.id.name.startsWith('exampleapp')) {
        details.belongsTo("aligned-platform:exampleapp:${details.id.version}")
    }
}
```
and uncomment:
```
resolutionRules files('alignment-rule.json')
```
to see how our plugin behaves.

**Comment by ljacomet** [comment]
The issue here is the holes in the versions.

Here are the relevant versions from this sample, the arrow indicating a dependency declaration, names and versions abbreviated for readability:
* `client:201` -> `common:225`
* `client:225` -> `common:249`

And here are the relevant steps:

1. At some point, the alignment logic tries to bump everything to version `225`, no problem, this succeeds.
2. `client:225` makes `common:249` enter the graph
3. Alignment logic tries to bump everything to version `249`. It works for `common` but fails for `client`.
4. No new versions are brought in
5. Graph resolution stops

The end result is similar to what you could have with unaligned versions due to some modules existing at a patch level and others not.

So as you can see, the main problem comes from the missing versions in that sequence.

The Gradle alignment logic only deals with version as seen in the graph. It does not transform any of the dependency declarations in a range.

**Comment by chali** [comment]
I thought this would be the same case like skewed versions of jackson. One library from the group has a patch version eg. `2.9.7.1` so rest of the group will be brought to `2.9.7` as close as possible. I know there is not `249` of the client but I was expecting the closest one which is `236`.

**Comment by ljacomet** [comment]
I understand where the confusion can be coming from, but we only guarantee alignment to the versions that are observed during resolution.

For example, and while #51 still needs a full answer, Jackson is a real mess and `jackson-databind:2.8.11.1` actually depends on `jackson-core:2.8.10` and not `jackson-core:2.8.11`. Assuming the other issue(s) #51 highlights are resolved, you would end up with one version a `2.8.11.1` and the rest at `2.8.10` because version `2.8.11` never appears in the graph at all.

**Comment by chali** [comment]
This is the expected behavior of the implementation`belongsTo`. I'm closing this one.
